### PR TITLE
GSdx: Improve captured screenshots naming

### LINKED ISSUE #1700: [Improvement] Faster sequential snapshots
***PCSX2 version:***
pcsx2-v1.5.0-dev-1711-g29d2292-windows-x86

***Description of the issue:***
I like to take a lot of screenshots of games, usually sequential snaps of a certain animation or cutscene. Problem is no matter how fast I press the snapshot key, there are a few seconds worth of delay between each generated screenshot. Meaning, that I can never get a quick sequential screenshot result.
I would be great if that delay was removed, so that if I press F8 four times in one second, I would get four screenshots of that specific time frame.

By the way, all settings are at default. As in, native resolution, no effects, no AA, etc.

***How to reproduce the issue:***
Just press F8 a bunch of times. 

***PC specifications:***
``(For example: CPU: Intel core i5 3470 GPU: none OS: Windows 7 64 Bit)``

### PR DESCRIPTION
**Pull request by ssakash**
**Summary of changes**:

* Prevent a conflict where it wasn't possible to capture multiple screenshots at a single second. (Refer commit message for further details)

Fixes #1700

### CODE REVIEW

**Comment by FlatOutPS2** [comment]
Woulnd't it be better/simpler to just add miliseconds?

**Comment by turtleli** [comment]
What's the actual goal? Multiple snapshots per second, or frame by frame snapshots?

**Comment by ssakash** [comment]
> Woulnd't it be better/simpler to just add milliseconds?

I'm not sure if adding milliseconds is the optimal approach, it would give you 3 extra digits at each captured file's name which seems a bit large instead of the current approach of adding just a single number alongside the file's name.

I could get the milliseconds, but I only know how to do it in WIN32 API, I'm not aware of any cross platform alternatives. Feel free to suggest one if you have any in your mind. Also do note that the current function for getting string of the local time (``strftime()``) is closely linked with the naming of the captured file, I'd prefer to not replace it with any other library function to save some trouble.

> What's the actual goal? Multiple snapshots per second, or frame by frame snapshots?

The goal is to capture multiple snapshots per second, I'm not sure what you mean by ``frame by frame snapshots``? Sounds confusing to me on what it actually means.

**Comment by turtleli** [comment]
Frame by frame snapshots - consecutive snapshots without missing frames?

**Comment by mirh** [comment]
Wouldn't that require "defining" what a *frame* is in the first place, which was one of the great problems in #444 ?

**Comment by turtleli** [comment]
No? I'm thinking of video capture but where the output is to separate image files without the resizing that the GS video recording does.

File: plugins/GSdx/GSRenderer.cpp [hunk 1]
```diff
@@ -552,17 +552,41 @@ void GSRenderer::VSync(int field)
 	}
 }
 
+static bool IsTimeEqual(tm t1, tm t2)
```
Comment (danilaml) [inline plugins/GSdx/GSRenderer.cpp]: any reason for pass by value?

Comment (ssakash) [inline plugins/GSdx/GSRenderer.cpp]: No specific reasons. Why should I pass by reference instead?

**Comment by ssakash** [comment]
> Frame by frame snapshots - consecutive snapshots without missing frames?

That would probably produce over 25-30 image captures per second, not sure if it would be feasible even with lowest compression factor. Also from my opinion, I don't think frame by frame capture would be useful. Even if you're making a GIF, merging 30+ images is quite time consuming, capturing something like 4 frames per second would be rather enough at most cases for making a decently smooth GIF. (Or, I guess so?)

**Comment by FlatOutPS2** [comment]
> I'm not sure if adding milliseconds is the optimal approach, it would give you 3 extra digits at each captured file's name which seems a bit large instead of the current approach of adding just a single number alongside the file's name.

True, but you are adding a space and (), so you're adding more characters anyway. :P

I do suggest changing it from `currentime (number)` to 'currenttime-number'. This way the sequential number snapsshots will added after the first entry in the snaps folder if the files are sorted by name or date.

And wouldn't a solution like this be simpler:

```
		static char prev_buff[16];
		static int n = 2;
		char buff[16];

		if(strftime(buff, sizeof(buff), "%Y%m%d%H%M%S", localtime(&t)))
		{
			if (!memcmp(buff, prev_buff, 16)) {
				m_snapshot = format("%s_%s (%d)", path.c_str(), buff, n++);
			}
			else {
				n = 2;
				m_snapshot = format("%s_%s", path.c_str(), buff);
			}
			memcpy(prev_buff, buff, sizeof(buff));
		}
	}
```

File: plugins/GSdx/GSRenderer.cpp (hunk 1 above)
Comment (danilaml) [inline plugins/GSdx/GSRenderer.cpp]: Well it kinda big (9 ints).

Comment (ssakash) [inline plugins/GSdx/GSRenderer.cpp]: I could remove this useless bool function if I follow [Flatout's idea](https://github.com/PCSX2/pcsx2/pull/1771#issuecomment-272634750). So yeah, I'll be doing that. ;)

**Comment by ssakash** [comment]
> I do suggest changing it from currentime (number) to 'currenttime-number'. This way the sequential number snapsshots will added after the first entry in the snaps folder if the files are sorted by name or date.

Sounds reasonable enough, I'll be using ``-number`` instead of ``(number)`` then. However, the latter seems more nice to be honest for the reasons I stated [earlier](https://github.com/PCSX2/pcsx2/issues/1700#issuecomment-272625104).

_Quoting my post_,
> it would be better to just add a (2),(3) next to the image's name to indicate that multiple images have been captured in a single second. **(Yeah, following the same format which Windows does when it wants to add a new file which has a similar name of an old file present in the current directory)** 

> And wouldn't a solution like this be simpler:

I would prefer to avoid the K&R coding style and the mem functions. :P

**Comment by turtleli** [comment]
> That would probably produce over 25-30 image captures per second, not sure if it would be feasible even with lowest compression factor. Also from my opinion, I don't think frame by frame capture would be useful. Even if you're making a GIF, merging 30+ images is quite time consuming, capturing something like 4 frames per second would be rather enough at most cases for making a decently smooth GIF. (Or, I guess so?)

Well, the idea is that you get to choose the images you want afterwards instead of mashing F8 and hoping you've captured the desired image. ;) But yeah, it's not mutually exclusive with mashing F8 to get multiple snapshots per second.

File: plugins/GSdx/GSRenderer.cpp [hunk 2]
```diff
@@ -557,12 +557,22 @@ bool GSRenderer::MakeSnapshot(const string& path)
 	if(m_snapshot.empty())
 	{
 		time_t t = time(NULL);
```
Comment (turtleli) [inline plugins/GSdx/GSRenderer.cpp]: Why not just compare time_t values?

File: plugins/GSdx/GSRenderer.cpp [hunk 3]
```diff
@@ -557,12 +557,22 @@ bool GSRenderer::MakeSnapshot(const string& path)
...
+		static int n = 2;
+		static char prev_snap[16];
+		char cur_time[16];
```
Comment (turtleli) [inline plugins/GSdx/GSRenderer.cpp]: Buffer overflow?

**Comment by ReyVGM** [comment]
Guys, don't overthink this. 

Just do what the Dolphin guys did with this very same issue was just add a (1), (2) after each image. For example:
Game 07:59.png
Game 08:00.png
Game 08:00 (1).png
Game 08:00 (2).png
Game 08:01.png
I believe that should be simple enough and would solve the issue.

I would advise against changing the naming from date/hour to sequential numbers because if you take a bunch of screenshots, and then delete an early screen, then when you take a new screen instead of continuing with the last number it will try to fill that slot you deleted and then you'll have non-chronological screenshots. With date/hour naming, that will never happen.

**Comment by Hirato** [comment]
I think it might be better if a _00 was tacked onto the end of the name, and then incremented to _01, _02, etc, in the event of collision.
it should account for people making any ridiculous number of screenshots per second, and it'll still be sorted.


That aside, isn't webm the go to format nowadays for people who want to record and share a little bit of footage?
You could just capture a few seconds of PCSX2 via obs and call it a day, and if you still want to make gifs from it, it might be a lot easier to step through that footage.

**Comment by ReyVGM** [comment]
You can't take pixel-perfect images from videos, much less make pixel perfect animated GIFs.

**Comment by MrCK1** [comment]
Isn't M-JPEG a bunch of individual images stitched together to make a video?

**Comment by mirh** [comment]
Pixel perfect means lossless. 
100 quality m-jpeg may possibly be visually indistinguishable from that, but if you have to convert and stuff on it, I guess at that point is just better to shoot uncompressed.

**Comment by gregory38** [comment]
On Linux, the GS video recording feature is implemented as consecutive frame dump (and people need to convert it to video). Maybe it can be extended. The PNG compression is done on separate threadS.

**Comment by ReyVGM** [comment]
Oh please don't mess with the screenshot button. Pressing F8 and getting a bunch of frame dump images is fine, but make it an option for people that want to make a sequence or animated GIF.

But for now, what needs to be improved is the naming of the screenshots that only allows one screen per second.

**Comment by gregory38** [comment]
IMHO, you need a gif option, that will transform the F8 as save N frames of data. It is much more clever than smashing F8. And there is already a linear naming.
```
     	m_out_dir = theApp.GetConfigS("capture_out_dir");
	std::string out_file = m_out_dir + format("/frame.%010d.png", m_frame);
```
However, I potentially broke the frame counter. Anyway, it can be improved without changing the standard snapshot behavior.

**Comment by ssakash** [comment]
So what's the conclusion on naming format of the screenshots? Should it be ``time (n)`` **(or)** ``time-n``
**(or)** ``time_n``?

**Comment by FlatOutPS2** [comment]
> So what's the conclusion on naming format of the screenshots? Should it be time (n) (or) time-n
(or) time_n?

It can also be `time_(n)` (but not `time-(n)`).

File: plugins/GSdx/GSRenderer.cpp [hunk 4]
```diff
@@ -556,13 +556,23 @@ bool GSRenderer::MakeSnapshot(const string& path)
 {
 	if(m_snapshot.empty())
 	{
-		time_t t = time(NULL);
+		time_t cur_time = time(nullptr);
+		static time_t prev_snap;
+		static int n = 2;
```
Comment (gregory38) [inline plugins/GSdx/GSRenderer.cpp]: Just a quick comment why the reset value is 2 and not 0 (or 1)

**Comment by turtleli** [comment]
> So what's the conclusion on naming format of the screenshots? Should it be `time (n)` (or) `time-n`
> (or) `time_n`?

I'd prefer one of the versions without spaces.

File: plugins/GSdx/GSRenderer.cpp (hunk 4 above)
Comment (ssakash) [inline plugins/GSdx/GSRenderer.cpp]: I'm not sure if you're asking me to add a comment on the source code (or) asking me to explain it to you right now?

Comment (gregory38) [inline plugins/GSdx/GSRenderer.cpp]: In the code. I guess to keep compatibility, there is no .1 image hence the .2 start.

**Comment by ssakash** [comment]
I went with ``time_(n)``, thanks for all the feedback!

**Pull request merged**